### PR TITLE
Release proposal

### PR DESCRIPTION
**Pull request by hanneshapke**
@casassg initial release notes. Just a proposal based on the TF Addons setup. 
Would love to connect regarding the `nighly` build setup.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -27,7 +27,11 @@ jobs:
     strategy:
       # Test for several versions of TFX in parallel
       matrix:
-        depconstraint: ["tfx~=1.0.0 tensorflow~=2.5.0", "tfx~=1.3.0 tensorflow~=2.6.0",]
+        depconstraint: [
+          "tfx~=1.0.0 tensorflow~=2.5.0",
```
Comment (casassg) [inline .github/workflows/ci.yml:31]: nit. Add note to keep in sync w versions in version.py

File: setup.py [hunk 2]
```diff
@@ -1,53 +1,64 @@
...
+    # Version
+    version = {}
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+    with open(os.path.join(base_dir, "tfx_addons", "version.py")) as fp:
+        exec(fp.read(), version)
+
+    project_name = "tfx-addons"
+    # TODO: Support nightly builds? Like https://github.com/tensorflow/addons/blob/d19c3efa5e4b9924e5f1e0b603fe0a22b8efcbfe/.github/workflows/release.yml#L65
+    if "--nightly" in sys.argv:
+        project_name = "tfx-addons-nightly"  # TODO: Confirm project names with rcrowe
```
Comment (casassg) [inline setup.py:25]: do we really think we will use nightly releases?

Comment (casassg) [inline setup.py:25]: do we really think we will use nightly releases?

File: tfx_addons/version.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Define TFX Addons version information."""
+
+# Required TFX version [min, max)
+INCLUSIVE_MIN_TFX_VERSION = "1.4.0"
+EXCLUSIVE_MAX_TFX_VERSION = "1.6.0"
```
Comment (casassg) [inline tfx_addons/version.py:19]: Add note to keep in sync w CI.

**Comment by rcrowe-google** [comment]
The Google team has reviewed and approved, but since this is still in draft I'll wait before sending formal approval.

File: setup.py (hunk 2 above)
Comment (hanneshapke) [reply setup.py]: removed

File: .github/workflows/ci.yml (hunk 1 above)
Comment (hanneshapke) [reply .github/workflows/ci.yml]: added

File: tfx_addons/__init__.py [hunk 4]
```diff
@@ -13,5 +13,3 @@
 # limitations under the License.
 # ==============================================================================
 """Init module for TFX."""
```
[the remark this answers is not in the document]
Comment (hanneshapke) [reply tfx_addons/__init__.py:15]: added back in

File: setup.py [hunk 5]
```diff
@@ -1,44 +1,50 @@
...
-import re
+import sys
 
 from setuptools import find_namespace_packages, setup
 
 
-def _get_version():
-  version_file = os.path.join(os.path.dirname(__file__),
-                              'tfx_addons/__init__.py')
-  with open(version_file, 'r') as fp:
```
Comment (pagillaprashanth) [inline setup.py:12]: with proper validation progress file can be deleted

**Comment by rcrowe-google** [comment]
Ready to merge?  I think you may be able to merge yourself.

File: tfx_addons/version.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# TODO: This license is not consistent with license used in the project.
+#       Delete the inconsistent license and above line and rerun pre-commit to insert a good license.
```
Comment (casassg) [inline tfx_addons/version.py:2]: make sure to update the template so this doesn't happen

File: setup.py [hunk 7]
```diff
@@ -1,44 +1,50 @@
...
+exclusive_max_tfx_version = version["EXCLUSIVE_MAX_TFX_VERSION"]
+
 NAME = "tfx-addons"
-# VERSION = .... Change the version in tfx_addons/__init__.py
 
 TESTS_REQUIRE = ["pytest", "pylint", "pre-commit", "isort", "yapf"]
 
 PKG_REQUIRES = {
     # Add dependencies here for your project. Avoid using install_requires.
     "mlmd_client": ["ml-pipelines-sdk>=1.0.0<2", "ml-metadata>=1.0.0<2"],
```
Comment (casassg) [inline setup.py:37]: add also the same constraints for ml-pipelines-sdk which is tfx but w less standard components

File: setup.py [hunk 8]
```diff
@@ -1,44 +1,50 @@
...
     "schema_curation": [
-        "tfx>=0.26.3<2.0.0",
+        "tfx>={},<{}".format(inclusive_min_tfx_version,
```
Comment (casassg) [inline setup.py:39]: nit: move string to a variable instead of subsituting several times

File: setup.py [hunk 9]
```diff
@@ -1,44 +1,50 @@
...
-      re.M,
-  )
-  if version_match:
-    return version_match.group(1)
-  else:
-    raise RuntimeError("Unable to find version string.")
+def get_last_commit_time() -> str:
+  string_time = os.getenv("NIGHTLY_TIME").replace('"', "")
+  return datetime.strptime(string_time,
+                           "%Y-%m-%dT%H:%M:%SZ").strftime("%Y%m%d%H%M%S")
```
Comment (casassg) [inline setup.py:14]: Is this still needed?

File: setup.py [hunk 10]
```diff
@@ -1,44 +1,50 @@
...
 
 
+def get_project_name_version():
+  # Version
+  extracted_version = {}
+  base_dir = os.path.dirname(os.path.abspath(__file__))
+  with open(os.path.join(base_dir, "tfx_addons", "version.py")) as fp:
+    exec(fp.read(), extracted_version)  # pylint: disable=exec-used
+
+  return PROJECT_NAME, extracted_version
```
Comment (casassg) [inline setup.py:24]: why are we returning project name?

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -26,10 +26,15 @@ jobs:
...
       # Test for several versions of TFX in parallel
+      # keep constraints in sync with version.py
       matrix:
         # NB(gcasassaez): TF 2.8 moves numpy to >=1.20 wich doesnt work for pyarrow<3
         # See: https://issues.apache.org/jira/browse/ARROW-11450
-        depconstraint: ["tfx~=1.0.0 tensorflow~=2.5.0", "tfx~=1.7.0 tensorflow~=2.8.0 pyarrow>=3.0.0"]
+        depconstraint: [
+          "tfx~=1.0.0 tensorflow~=2.5.0"
+          "tfx~=1.4.0 tensorflow>=2.6.0",
+          "tfx~=1.7.0 tensorflow>=2.8.0 pyarrow>=3.0.0",
```
Comment (casassg) [inline .github/workflows/ci.yml:36]: Let's use ~= better

**Comment by hanneshapke** [comment]
@casassg Thank you for review. Ready from my end.

File: .github/workflows/publish-to-test-pypi.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+name: Release TFX Addons package to PyPI and TestPyPI
+
+on:
+  workflow_run:
+    workflows: ["TFX Addons package CI", "Lint"]
+    branches: [main]
+    types:
+      - completed
```
Comment (casassg) [inline .github/workflows/publish-to-test-pypi.yml:8]: What happens if we want to backport a commit to a previous version (aka patch release). We should trigger on any r* branches as well if possible

**Review by casassg — APPROVED** [review]

**Review by rcrowe-google — APPROVED** [review]
Thanks!

**Pull request merged**